cmd/bamTagToReadGroup/bamTagToReadGroup.go

```diff
@@ -0,0 +1,141 @@
+		usage()
+		log.Fatalln("ERROR: must provide a bam file with -i")
+	}
+
+	if *tagId == "" || len(*tagId) != 2 {
+		usage()
+		log.Fatalln("ERROR: -tagId must be 2 characters")
+	}
+
+	if *tagValuesFile == "" {
+		usage()
+		log.Fatalln("ERROR: must provide a tag values file with -tagValues")
+	}
+
+	bamTagToReadGroup(*infile, *outfile, *tagId, *tagValuesFile)
+}
+
+func bamTagToReadGroup(infile, outfile, tagId, tagValuesFile string) {
+	tagValues := fileio.Read(tagValuesFile)
+	tagValueMap := make(map[string]bool)
+	for i := range tagValues {
+		tagValueMap[tagValues[i]] = true
+	}
+
+	inputRecordChan, header := sam.GoReadToChan(infile)
+	addTagsToHeader(&header, tagValues)
+
+	out := fileio.EasyCreate(outfile)
+	defer safeclose(out)
+	bw := sam.NewBamWriter(out, header)
+	defer safeclose(bw)
+
+	// spawn goroutine to write records after being updated
+	writeChan := make(chan sam.Sam, 1000)
+	wg := new(sync.WaitGroup)
+	wg.Add(1)
+	go func(writeChan <-chan sam.Sam, bw *sam.BamWriter, wg *sync.WaitGroup) {
+		for rec := range writeChan {
+			sam.WriteToBamFileHandle(bw, rec, 0)
+		}
+		wg.Done()
+	}(writeChan, bw, wg)
+
+	// read records and update read groups as we go
+	for rec := range inputRecordChan {
+		updateRecord(&rec, tagId, tagValueMap)
+		writeChan <- rec
+	}
+	close(writeChan)
+
+	// wait for goroutine to finish writing before return on main thread
+	wg.Wait()
+}
+
+func safeclose(f io.Closer) {
+	err := f.Close()
+	exception.PanicOnErr(err)
+}
+
+// addTagsToHeader adds each element in tagValues as a read group in the header
+// format is `@RG	ID:$TAGVALUE	SM:$TAGVALUE	LB:$TAGVALUE`
+func addTagsToHeader(header *sam.Header, tagValues []string) {
+	newHeaderText := make([]string, 0, len(header.Text)+len(tagValues))
+	for i := range header.Text {
+		if !strings.HasPrefix(header.Text[i], "@RG") {
+			newHeaderText = append(newHeaderText, header.Text[i])
+		}
+	}
+	for i := range tagValues {
+		newHeaderText = append(newHeaderText, fmt.Sprintf("@RG\tID:%s\tSM:%s\tLB:%s", tagValues[i], tagValues[i], tagValues[i]))
+	}
+	header.Text = newHeaderText
+}
+
+func updateRecord(rec *sam.Sam, tagId string, tagValueMap map[string]bool) {
+	var err error
+
+	// parse the extra field where tags are since we are coming from bam
+	err = sam.ParseExtra(rec)
+	exception.PanicOnErr(err)
+
+	splitExtra := strings.Split(rec.Extra, "\t")
+	var newExtra []string = make([]string, 0, len(splitExtra)+1)
+
+	var tagValue string
+
+	// first remove any exiting read group and pull the tag value at the same time
+	for i := range splitExtra {
+		if !strings.HasPrefix(splitExtra[i], "RG:Z:") {
+			newExtra = append(newExtra, splitExtra[i])
+		}
+
+		if strings.HasPrefix(splitExtra[i], tagId+":") {
+			tagValue = splitExtra[i][5:] // skips characters defining tag (e.g. CB:Z:TagValue)
+		}
+	}
+
+	// add read group if tag value is present in input values file
+	if tagValueMap[tagValue] {
+		newExtra = append(newExtra, fmt.Sprintf("RG:Z:%s", tagValue))
```

Comment:
Maybe it make sense to refactor the `addTagsToHeader` to something like `CreateTag` or `AddTag` to return a string representation of the new tag w/ its value so you can reuse the same code for both the header and updating the records.

Reply:
Agreed. Playing with tags is a bit error-prone when starting from bams since changes to the text wont propagate to the output unless you run ParseExtra first. Will probably save us some headaches in the future by having builtin functions for this. I am going to implement these now.